<?xml version="1.0"?>
<launch>

  <arg name="mav_name" default="iris" />
  <arg name="x" default="0.0"/>
  <arg name="y" default="-2.0"/>
  <arg name="z" default="0.31"/>
  <arg name="Y" default="-1.570796"/>
  <arg name="map_only" default="false"/>
  <arg name="map_filename" default=""/>
  <arg name="output" default="screen"/>
  <arg name="auto_enable_by_slam" default="false"/>
  <arg name="auto_pilot_type" default="default"/>
  <arg name="save_traj" default="false"/>

  <!-- beacon -->
  <node name="beacon" pkg="mav_tunnel_nav" type="beacon" output="$(arg output)">
    <!-- robot's name used for interactions with other robots -->
    <param name="robot_name" value="$(arg mav_name)"/>
  </node>

  <!-- Convert cmd_vel into pose command so as to provide the same interface -->
  <node name="cmdvel_transform" pkg="mav_tunnel_nav" type="cmdvel_converter.py" args="odometry_sensor1/pose cmd_vel command/pose" output="$(arg output)"/>
  <!-- <node name="cmdvel_transform" pkg="mav_tunnel_nav" type="cmdvel_converter.py" args="imu_odom/pose cmd_vel command/pose"/> -->

  <!-- Autonomous control based on ranging data -->
  <node name="auto_pilot" pkg="mav_tunnel_nav" type="auto_pilot" output="$(arg output)">
    <param name="enable_topic" value="enable"/>
    <param name="update_freq" value="5"/>
    <!-- refer to xacro for range settings -->
    <param name="range_list"
      value = "range_front  0.11     0 -0.01 0 0   0
               range_rfront 0.11 -0.06 -0.01 0 0 -45
               range_lfront 0.11  0.06 -0.01 0 0  45

               range_right  0 -0.06 0.01 0 0 -90
               range_left   0  0.06 0.01 0 0  90

               range_rear   -0.12     0 -0.01 0 0  180
               range_rrear  -0.12 -0.06 -0.01 0 0 -135
               range_lrear  -0.12  0.06 -0.01 0 0  135

               range_up          0 0 0.055 0 -90   0
               range_ufront  0.037 0 0.055 0 -45   0
               range_urear  -0.037 0 0.055 0 -45 180

               range_down        0 0 -0.085 0 90   0
               range_dfront  0.037 0 -0.085 0 45   0
               range_drear  -0.037 0 -0.085 0 45 180"
      />

    <!-- These must be the same as those in the xacro file of Iris -->
    <!-- see iris_depthcam.xacro -->
    <param name="range_max" value="3.0"/>
    <param name="range_min" value="0.01"/>

    <!-- lower, more likely to steer -->
    <param name="STEER_THRESH" value="0.05"/>
    <param name="MIDDLE_THRESH" value="0.1"/>
    <!-- condition1: higher, more likely turn left -->
    <param name="TURN_THRESH1" value="1.2"/>
    <!-- condition2: (if condition1 unmet) higher, more likely turn -->
    <param name="TURN_THRESH2" value="0.9"/>
    <!-- condition3: (if condition2 met) higher, more likely turn right -->
    <param name="TURN_THRESH3" value="1.1"/>
    <param name="ALT_THRESH" value="0.05"/>
    <param name="OBS_THRESH" value="0.2"/>

    <param name="straight_rate" value="0.3"/>
    <param name="steering_yaw_rate" value="1.0"/>
    <param name="mid_open" value="0.8"/>
    <param name="middle_line_rate" value="2.0"/>
    <param name="turn_yaw_rate" value="1.0"/>
    <param name="alt_rate" value="2.0"/>
    <param name="obs_rate" value="0.15"/>
    <param name="alt_open" value="0.8"/>

    <param name="max_linear_x" value="0.3"/>
    <param name="max_linear_y" value="0.3"/>
    <param name="max_linear_z" value="0.5"/>

    <param name="distance_to_neighbor" value="15.0"/>
    <param name="rel_pose_expiration" value="3.0" />

    <param name="beacon_down_topic" value="beacon_down_topic"/>
    <param name="base_station_name" value="base"/>

    <param name="auto_pilot_type" value="$(arg auto_pilot_type)"/>
  </node>

  <!-- Sends motor commands based given commanded rates for roll/pitch/yaw and thrust -->
  <!-- <node name="roll_pitch_yawrate_thrust_controller_node" pkg="rotors_control" type="roll_pitch_yawrate_thrust_controller_node">
    <rosparam command="load" file="$(find rotors_gazebo)/resource/roll_pitch_yawrate_thrust_controller_iris.yaml" />
    <rosparam command="load" file="$(find rotors_gazebo)/resource/iris.yaml" />
    <remap from="odometry" to="imu_odom/odometry" />
  </node> -->
  <!-- Spawn a position controller -->
  <node name="lee_position_controller_node" pkg="rotors_control" type="lee_position_controller_node" output="$(arg output)">
    <rosparam command="load" file="$(find rotors_gazebo)/resource/lee_controller_iris.yaml" />
    <rosparam command="load" file="$(find rotors_gazebo)/resource/iris.yaml" />
    <remap from="odometry" to="odometry_sensor1/odometry" />
    <!-- <remap from="odometry" to="imu_odom/odometry" /> -->
  </node>

  <!-- imu odometry (optional) -->
  <!--
  <node name="imu_odom" pkg="mav_tunnel_nav" type="imu_odom" output="$(arg output)">
    <param name="init_Y" value="$(arg Y)"/>
    <param name="odom_reset_topic" value="odom_reset"/>
    <param name="imu_topic" value="imu"/>
    <param name="odom_topic" value="imu_odom/odometry"/>
    <param name="odom_pose_topic" value="imu_odom/pose"/>
    <param name="world_frame_id" value="world"/>
    <param name="odom_frame_id" value="$(arg mav_name)/odom"/>
  </node>-->

  <!-- rbpf -->
  <group unless="$(arg map_only)">
    <node name="rbpf" pkg="mav_tunnel_nav" type="rbpf" output="$(arg output)">
      <!-- robot's name used for interactions with other robots -->
      <param name="robot_name" value="$(arg mav_name)"/>

      <!-- <param name="odom_topic" value="imu_odom/odometry"/> -->
      <param name="odom_topic" value="odometry_sensor1/odometry"/>
      <param name="odom_reset_topic" value="odom_reset"/>
      <param name="pc_topic" value="camera_ir/camera/depth/points"/>
      <param name="world_frame_id" value="world"/>
      <param name="robot_frame_id" value="$(arg mav_name)/base_link"/>

      <param name="n_particles" value="50"/>
      <param name="update_freq" value="1"/>

      <param name="depth_cam_pc_downsample" value="1000"/>

      <param name="init_x" value="$(arg x)"/>
      <param name="init_y" value="$(arg y)"/>
      <param name="init_z" value="$(arg z)"/>
      <param name="init_Y" value="$(arg Y)"/>
      <param name="map_resol" value="0.2"/>
      <param name="map_probHit" value="0.8"/>
      <param name="map_probMiss" value="0.2"/>
      <param name="map_threshMin" value="0.12"/>
      <param name="map_threshMax" value="0.97"/>

      <param name="motion_noise_lin_sigma" value="0.005"/>
      <param name="motion_noise_rot_sigma" value="0.01"/>

      <param name="sensor_noise_range_sigma" value="1000.0"/>
      <param name="sensor_noise_depth_sigma" value="10000.0"/>

      <param name="t_pose_adjust" value="1.5"/>
      <param name="t_only_mapping" value="2.0"/>
      <param name="mapping_interval" value="9"/>
      <param name="publish_interval" value="0"/>
      <param name="vismap_interval" value="3"/>
      <param name="compress_interval" value="7"/>
      <param name="enable_indivLoc" value="true"/>
      <param name="seed_indivloc" value="1500"/>
      <!-- <param name="locdata_interval" value="1000"/> -->

      <param name="octomap_topic" value="octomap"/>

      <!-- refer to xacro for range settings -->
      <param name="range_list"
        value = "range_front  0.11     0 -0.01 0 0   0
                 range_rfront 0.11 -0.06 -0.01 0 0 -45
                 range_lfront 0.11  0.06 -0.01 0 0  45

                 range_right  0 -0.06 0.01 0 0 -90
                 range_left   0  0.06 0.01 0 0  90

                 range_rear   -0.12     0 -0.01 0 0  180
                 range_rrear  -0.12 -0.06 -0.01 0 0 -135
                 range_lrear  -0.12  0.06 -0.01 0 0  135

                 range_up          0 0 0.055 0 -90   0
                 range_ufront  0.037 0 0.055 0 -45   0
                 range_urear  -0.037 0 0.055 0 -45 180

                 range_down        0 0 -0.085 0 90   0
                 range_dfront  0.037 0 -0.085 0 45   0
                 range_drear  -0.037 0 -0.085 0 45 180"
        />
      <param name="range_max" value="1.5"/>
      <param name="range_min" value="0.01"/>

      <!-- parameters for cooperative localization -->
      <param name="enable_cooploc" value="true"/>
      <param name="enable_conservative" value="true"/>
      <param name="conserv_omega" value="0.6"/>
      <param name="sigma_kde" value="0.2"/>
      <param name="Nref" value="20"/>
      <param name="seed_cooploc" value="1000"/>
      <param name="sigmaLocR" value="0.5"/>
      <param name="sigmaLocT" value="0.05"/>
      <param name="gl_eval_cons" value="2"/>
      <param name="ml_eval_cons" value="2"/>

      <param name="beacon_lifetime" value="1.0"/>
      <param name="cooploc_phase" value="0.4"/>
      <param name="syncinit_timeout" value="0.1"/>

      <param name="auto_enable_by_slam" value="$(arg auto_enable_by_slam)"/>

      <!-- parameters for segmentation -->
      <param name="enable_segmentation" value="true"/>
      <param name="init_seg_phase" value="3.0"/>
      <param name="next_seg_thresh" value="10.0"/>
      <param name="enable_clr4seg" value="true"/>
      <param name="n_submaps" value="1"/>

      <!-- for evaluation -->
      <param name="save_traj" value="$(arg save_traj)"/>
    </node>
  </group>
  <!-- mapping only (using the ground truth poses) -->
  <group if="$(arg map_only)">
    <node name="map_only" pkg="mav_tunnel_nav" type="map_only" output="$(arg output)">
      <param name="pc_topic" value="camera_ir/camera/depth/points"/>
      <param name="world_frame_id" value="world"/>
      <param name="robot_frame_id" value="$(arg mav_name)/base_link"/>

      <param name="update_freq" value="2"/>

      <param name="depth_cam_pc_downsample" value="1000"/>

      <param name="init_x" value="$(arg x)"/>
      <param name="init_y" value="$(arg y)"/>
      <param name="init_z" value="$(arg z)"/>
      <param name="init_Y" value="$(arg Y)"/>
      <param name="map_resol" value="0.2"/>
      <param name="map_probHit" value="0.8"/>
      <param name="map_probMiss" value="0.2"/>
      <param name="map_threshMin" value="0.12"/>
      <param name="map_threshMax" value="0.97"/>

      <param name="map_filename" value="$(arg map_filename)"/>

      <param name="mapping_interval" value="3"/>
      <param name="publish_interval" value="0"/>
      <param name="vismap_interval" value="3"/>
      <param name="compress_interval" value="7"/>

      <param name="octomap_topic" value="octomap"/>
    </node>
  </group>

  <!-- map saver -->
  <node name="map_saver" pkg="mav_tunnel_nav" type="map_saver" output="$(arg output)">
    <param name="robot_name" value="$(arg mav_name)"/>
    <param name="filename" value="octomap.bt"/>
    <param name="octomap_topic" value="octomap"/>
    <param name="ack_topic" value="submap_ack_down_topic"/>
    <param name="savemap_topic" value="savemap"/>
  </node>

  <!-- Given robot_description parameter and joint states, publish TF frames -->
  <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" output="$(arg output)"/>
  <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher" output="$(arg output)"/>

</launch>
